src/box/vy_scheduler.c

```diff
@@ -350,13 +350,28 @@ vy_worker_pool_start(struct vy_worker_pool *pool)
 	}
 }
 
+/**
+ * Notify worker threads of shutdown. The worker thread task in progress will
+ * be cancelled right after it is yield.
+ */
+static void
+vy_worker_pool_start_shutdown(struct vy_worker_pool *pool)
+{
+	if (pool->workers == NULL)
+		return;
+	for (int i = 0; i < pool->size; i++)
+		cbus_stop_loop(&pool->workers[i].worker_pipe);
```

Comment:
Please try to make it consistent with `vy_run_env_stop_readers`, without two phases.

Reply:
Done. Now we finish worker threads during shutdown.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/vy_scheduler.c b/src/box/vy_scheduler.c
index 3d3e344f6..0f39deb46 100644
--- a/src/box/vy_scheduler.c
+++ b/src/box/vy_scheduler.c
@@ -355,22 +355,15 @@ vy_worker_pool_start(struct vy_worker_pool *pool)
  * be cancelled right after it is yield.
  */
 static void
-vy_worker_pool_start_shutdown(struct vy_worker_pool *pool)
+vy_worker_pool_shutdown(struct vy_worker_pool *pool)
 {
 	if (pool->workers == NULL)
 		return;
-	for (int i = 0; i < pool->size; i++)
-		cbus_stop_loop(&pool->workers[i].worker_pipe);
-}
-
-static void
-vy_worker_pool_stop(struct vy_worker_pool *pool)
-{
-	assert(pool->workers != NULL);
 	for (int i = 0; i < pool->size; i++) {
 		struct vy_worker *worker = &pool->workers[i];
+		cbus_stop_loop(&worker->worker_pipe);
 		cpipe_destroy(&worker->worker_pipe);
-		if (cord_join(&worker->cord) != 0)
+		if (cord_cojoin(&worker->cord) != 0)
 			panic_syserror("failed to join vinyl worker thread");
 	}
 	free(pool->workers);
@@ -386,13 +379,6 @@ vy_worker_pool_create(struct vy_worker_pool *pool, const char *name, int size)
 	stailq_create(&pool->idle_workers);
 }
 
-static void
-vy_worker_pool_destroy(struct vy_worker_pool *pool)
-{
-	if (pool->workers != NULL)
-		vy_worker_pool_stop(pool);
-}
-
 /**
  * Get an idle worker from a pool.
  */
@@ -491,9 +477,8 @@ vy_scheduler_finish_tasks(struct vy_scheduler *scheduler, int *tasks_done,
 void
 vy_scheduler_shutdown(struct vy_scheduler *scheduler)
 {
-	vy_worker_pool_start_shutdown(&scheduler->dump_pool);
-	vy_worker_pool_start_shutdown(&scheduler->compaction_pool);
-
+	vy_worker_pool_shutdown(&scheduler->dump_pool);
+	vy_worker_pool_shutdown(&scheduler->compaction_pool);
 	fiber_cancel(scheduler->scheduler_fiber);
 	fiber_join(scheduler->scheduler_fiber);
 	scheduler->scheduler_fiber = NULL;
@@ -510,8 +495,6 @@ vy_scheduler_shutdown(struct vy_scheduler *scheduler)
 void
 vy_scheduler_destroy(struct vy_scheduler *scheduler)
 {
-	vy_worker_pool_destroy(&scheduler->dump_pool);
-	vy_worker_pool_destroy(&scheduler->compaction_pool);
 	diag_destroy(&scheduler->diag);
 	fiber_cond_destroy(&scheduler->dump_cond);
 	fiber_cond_destroy(&scheduler->scheduler_cond);

```

</details>